File: java/revenj-core/src/main/java/org/revenj/RevenjSystemState.java
package org.revenj;

import org.revenj.extensibility.Container;
import org.revenj.extensibility.SystemState;
import rx.Observable;
import rx.subjects.PublishSubject;

class RevenjSystemState implements SystemState {

	private boolean systemBooting = true;
	private boolean systemReady = false;
	private final PublishSubject<SystemEvent> changeSubject = PublishSubject.create();
	private final PublishSubject<Container> startupSubject = PublishSubject.create();
	private final Observable<SystemEvent> changeEvents;
	private final Observable<Container> startupEvents;

	public RevenjSystemState() {
		changeEvents = changeSubject.asObservable();
		startupEvents = startupSubject.asObservable();
	}

	@Override
	public boolean isBooting() {
		return systemBooting;
	}

	@Override
	public boolean isReady() {
		return systemReady;
	}

	void started(Container container) {
		systemBooting = false;
		systemReady = true;
		startupSubject.onNext(container);
	}

	@Override
	public Observable<Container> ready() {
		return startupEvents;
	}

	@Override
	public Observable<SystemEvent> change() {
		return changeEvents;
	}

	@Override
	public void notify(SystemEvent value) {
		changeSubject.onNext(value);
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/RevenjSystemState.scala
package net.revenj

import monix.reactive.{MulticastStrategy, Observable}
import monix.reactive.subjects.ConcurrentSubject
import net.revenj.extensibility.{Container, SystemState}

private[revenj] class RevenjSystemState extends SystemState {
  private var systemBooting = true
  private var systemReady = false
  private val changeSubject = ConcurrentSubject[SystemState.SystemEvent](MulticastStrategy.publish)(monix.execution.Scheduler.Implicits.global)
  private val startupSubject = ConcurrentSubject[Container](MulticastStrategy.publish)(monix.execution.Scheduler.Implicits.global)
  private val changeEvents = changeSubject.map(identity)
  private val startupEvents = startupSubject.map(identity)

  def isBooting: Boolean = systemBooting

  def isReady: Boolean = systemReady

  private[revenj] def started(container: Container): Unit = {
    systemBooting = false
    systemReady = true
    startupSubject.onNext(container)
  }

  def ready: Observable[Container] = startupEvents

  def change: Observable[SystemState.SystemEvent] = changeEvents

  def notify(value: SystemState.SystemEvent): Unit = {
    changeSubject.onNext(value)
  }
}
